lib/charms/loki_k8s/v0/charm_logging.py

```diff
@@ -0,0 +1,463 @@
+# The unique Charmhub library identifier, never change it
+LIBID = "52ee6051f4e54aedaa60aa04134d1a6d"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+PYDEPS = []
+
+logger = logging.getLogger("charm_logging")
+_GetterType = Union[Callable[[CharmBase], Optional[str]], property]
+CHARM_LOGGING_ENABLED = "CHARM_LOGGING_ENABLED"
+
+
+# from https://github.com/GreyZmeem/python-logging-loki, which seems to be dead
+class LokiEmitter:
+    """Base Loki emitter class."""
+
+    #: Success HTTP status code from Loki API.
+    success_response_code: int = 204
+
+    #: Label name indicating logging level.
+    level_tag: str = "severity"
+    #: Label name indicating logger name.
+    logger_tag: str = "logger"
+
+    #: String contains chars that can be used in label names in LogQL.
+    label_allowed_chars: str = "".join((string.ascii_letters, string.digits, "_"))
+    #: A list of pairs of characters to replace in the label name.
+    label_replace_with: Tuple[Tuple[str, str], ...] = (
+        ("'", ""),
+        ('"', ""),
+        (" ", "_"),
+        (".", "_"),
+        ("-", "_"),
+    )
+
+    def __init__(self, url: str, tags: Optional[dict] = None, cert: Optional[str] = None):
+        """Create new Loki emitter.
+
+        Arguments:
+            url: Endpoint used to send log entries to Loki (e.g. `https://my-loki-instance/loki/api/v1/push`).
+            tags: Default tags added to every log record.
+            cert: Absolute path to a ca cert for TLS authentication.
+
+        """
+        #: Tags that will be added to all records handled by this handler.
+        self.tags = tags or {}
+        #: Loki JSON push endpoint (e.g `http://127.0.0.1/loki/api/v1/push`)
+        self.url = url
+        #: Optional cert for TLS auth
+        self.cert = cert
+
+        self._session: Optional[requests.Session] = None
+
+    def __call__(self, record: logging.LogRecord, line: str):
+        """Send log record to Loki."""
+        payload = self.build_payload(record, line)
+        resp = self.session.post(self.url, json=payload, timeout=5)
+        if resp.status_code != self.success_response_code:
+            raise ValueError(
+                "Unexpected Loki API response status code: {0}".format(resp.status_code)
+            )
+
+    def build_payload(self, record: logging.LogRecord, line) -> dict:
+        """Build JSON payload with a log entry."""
+        labels = self.build_tags(record)
+        ns = 1e9
+        ts = str(int(time.time() * ns))
+        stream = {
+            "stream": labels,
+            "values": [[ts, line]],
+        }
+        return {"streams": [stream]}
+
+    @property
+    def session(self) -> requests.Session:
+        """Create HTTP(s) session."""
+        if self._session is None:
+            self._session = requests.Session()
```

Comment:
Iirc, requests is a heavy dep to build from source.
Would it be easy to replace with urllib?

Reply:
good point. If it's a little work I'll do this otherwise we can have a follow-up iteration to thin it down

Reply:
fixed